lib/df-detection.c

```diff
@@ -0,0 +1,133 @@
+#include "linux/df-detection.h"
+#include <asm/syscall.h>
+#include <linux/context_tracking.h>
+#include <linux/entry-common.h>
+#include <linux/kernel.h>
+#include <linux/livepatch.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+void add_address(const void *addr, size_t len, unsigned long caller)
+{
+	if (current->addresses == NULL || current->pairs == NULL)
+		return;
+	if (current->num_read >= current->sz &&
+	    reallocate_extra_memory(current->sz, DF_MAX_RECORDS)) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		detect_intersection();
+		current->num_read++;
+	}
+}
+
+void start_system_call(long syscall)
+{
+	if (syscall == 54 || syscall == 165 || syscall == 16)
```

Comment:
I think we'd better keep this logic in report(), or maybe introduce a special function that tells us whether to print the report or not. In the future the filtering rules will become more complex and will probably depend on other data that's not available to start_system_call().